### PR TITLE
XML documentation for Normalizer

### PR DESCRIPTION
**Pull request by singlis**
Tracked in #3204

### CODE REVIEW

File: src/Microsoft.ML.Data/Transforms/Normalizer.cs [hunk 1]
```diff
@@ -26,6 +26,36 @@
...
+    /// <summary>
+    /// <see cref="IEstimator{TTransformer}"/> for the <see cref="NormalizingTransformer"/>.
+    /// </summary>
+    /// <remarks>
+    /// <format type="text/markdown"><![CDATA[
+    ///
+    /// ###  Estimator Characteristics
+    /// |  |  |
+    /// | -- | -- |
+    /// | Does this estimator need to look at the data to train its parameters? | No |
```
Comment (natke) [inline src/Microsoft.ML.Data/Transforms/Normalizer.cs]: Is this correct? Doesn't it have to look at the data to find the mean, variance, min, max etc?

File: src/Microsoft.ML.Data/Transforms/Normalizer.cs [hunk 2]
```diff
@@ -26,6 +26,36 @@
...
+    /// | Input column data type | <xref:System.Single> or <xref:System.Double> |
+    /// | Output column data type | The same as the data type in the input column |
+    ///
+    /// The resulting NormalizingEstimator will normalize the data in one of the following ways based upon how it was created:
+    /// * Min Max - Linear rescale such that minimum and maximum values are mapped between -1 and 1.
+    /// * Mean Variance - Rescale to unit variance and, optionally, zero mean.
+    /// * Log Mean Variance - Rescale to unit variance on the log scale.
+    /// * Binning - Bucketize and then rescale to between -1 and 1.
```
Comment (natke) [inline src/Microsoft.ML.Data/Transforms/Normalizer.cs]: Is this correct? Isn't it between 0 and 1?

File: src/Microsoft.ML.Data/Transforms/Normalizer.cs [hunk 3]
```diff
@@ -26,6 +26,36 @@
...
+    ///
+    /// ###  Estimator Characteristics
+    /// |  |  |
+    /// | -- | -- |
+    /// | Does this estimator need to look at the data to train its parameters? | No |
+    /// | Input column data type | <xref:System.Single> or <xref:System.Double> |
+    /// | Output column data type | The same as the data type in the input column |
+    ///
+    /// The resulting NormalizingEstimator will normalize the data in one of the following ways based upon how it was created:
+    /// * Min Max - Linear rescale such that minimum and maximum values are mapped between -1 and 1.
```
Comment (shmoradims) [inline src/Microsoft.ML.Data/Transforms/Normalizer.cs]: >between -1 and 1 [](start = 83, length = 16)

let's use ranges: [-1,1] or even better with latex: $[-1,1]$

File: src/Microsoft.ML.Data/Transforms/Normalizer.cs [hunk 4]
```diff
@@ -26,6 +26,36 @@
...
+    /// <see cref="IEstimator{TTransformer}"/> for the <see cref="NormalizingTransformer"/>.
+    /// </summary>
+    /// <remarks>
+    /// <format type="text/markdown"><![CDATA[
+    ///
+    /// ###  Estimator Characteristics
+    /// |  |  |
+    /// | -- | -- |
+    /// | Does this estimator need to look at the data to train its parameters? | No |
+    /// | Input column data type | <xref:System.Single> or <xref:System.Double> |
```
Comment (shmoradims) [inline src/Microsoft.ML.Data/Transforms/Normalizer.cs]: >  [](start = 79, length = 1)

I think numeric vectors are also valid, aren't they? If so please change here and also the params. you can write it like this:

<xref:System.Single>, <xref:System.Double>, or a known-sized vector of those types.

File: src/Microsoft.ML.Transforms/NormalizerCatalog.cs [hunk 5]
```diff
@@ -31,11 +31,12 @@ public static class NormalizationCatalog
         }
 
         /// <summary>
-        /// It normalizes the data based on the observed minimum and maximum values of the data.
+        /// Create a <see cref="NormalizingEstimator"/>, which normalizes based on the observed minimum and maximum values of the data.
         /// </summary>
         /// <param name="catalog">The transform catalog</param>
         /// <param name="outputColumnName">Name of the column resulting from the transformation of <paramref name="inputColumnName"/>.</param>
```
Comment (shmoradims) [inline src/Microsoft.ML.Transforms/NormalizerCatalog.cs]: >. [](start = 133, length = 1)

Add: Output data type will be the same as input data type.

File: src/Microsoft.ML.Data/Transforms/Normalizer.cs (hunk 4 above)
Comment (singlis) [reply src/Microsoft.ML.Data/Transforms/Normalizer.cs]: Its only supports single or double:
[https://github.com/dotnet/machinelearning/blob/ea5c095f90cced3bd2c67da7289fe1be532f9ba4/src/Microsoft.ML.Data/Transforms/Normalizer.cs#L271](https://github.com/dotnet/machinelearning/blob/ea5c095f90cced3bd2c67da7289fe1be532f9ba4/src/Microsoft.ML.Data/Transforms/Normalizer.cs#L271)


---
In reply to: [276994571](https://github.com/dotnet/machinelearning/pull/3432#discussion_r276994571) [](ancestors = 276994571)

File: src/Microsoft.ML.Data/Transforms/Normalizer.cs (hunk 1 above)
Comment (singlis) [reply src/Microsoft.ML.Data/Transforms/Normalizer.cs]: It should be yes, thanks for mentioning. I will correct this.

---
In reply to: [276894215](https://github.com/dotnet/machinelearning/pull/3432#discussion_r276894215) [](ancestors = 276894215)

File: src/Microsoft.ML.Data/Transforms/Normalizer.cs (hunk 2 above)
Comment (singlis) [reply src/Microsoft.ML.Data/Transforms/Normalizer.cs]: it looks like -1 and 1. See the sample:
[https://github.com/dotnet/machinelearning/blob/ea5c095f90cced3bd2c67da7289fe1be532f9ba4/docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeBinning.cs#L58](https://github.com/dotnet/machinelearning/blob/ea5c095f90cced3bd2c67da7289fe1be532f9ba4/docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/NormalizeBinning.cs#L58)


---
In reply to: [276894276](https://github.com/dotnet/machinelearning/pull/3432#discussion_r276894276) [](ancestors = 276894276)

File: src/Microsoft.ML.Data/Transforms/Normalizer.cs (hunk 4 above)
Comment (shmoradims) [reply src/Microsoft.ML.Data/Transforms/Normalizer.cs]: i think known-sized vectors of numerics are supported by normalizer. that line you're referencing is also true if the vector contains only numeric values. could you please confirm with Ivan?

I'm certain that I've normalized the feature **vector** right before trainer

---
In reply to: [277045796](https://github.com/dotnet/machinelearning/pull/3432#discussion_r277045796) [](ancestors = 277045796,276994571)

Comment (Ivanidzo4ka) [reply src/Microsoft.ML.Data/Transforms/Normalizer.cs]: Look on line 298 in this file.
It's either scalar or vector of known-size of float or double

---
In reply to: [277069180](https://github.com/dotnet/machinelearning/pull/3432#discussion_r277069180) [](ancestors = 277069180,277045796,276994571)

File: src/Microsoft.ML.Data/Transforms/Normalizer.cs [hunk 6]
```diff
@@ -26,6 +26,35 @@
...
+    ///
+    /// ###  Estimator Characteristics
+    /// |  |  |
+    /// | -- | -- |
+    /// | Does this estimator need to look at the data to train its parameters? | Yes |
+    /// | Input column data type | <xref:System.Single> or<xref:System.Double> |
+    /// | Output column data type | The same data type as the input column |
+    ///
+    /// The resulting NormalizingEstimator will normalize the data in one of the following ways based upon how it was created:
+    /// * Min Max - Linear rescale such that the values are mapped to the $[-1,1]$ with the minimum and maximum values being mapped to -1 and 1 respectively.
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Data/Transforms/Normalizer.cs]: >-1,1] [](start = 76, length = 5)

If we not talking about fixZero option all normalizers try to map to [0,1] range.
FixZero is special beast....
Not sure it's worth to specify range...

File: src/Microsoft.ML.Data/Transforms/Normalizer.cs [hunk 7]
```diff
@@ -26,6 +26,35 @@
...
+    ///
+    /// To create this estimator use one of the following:
+    /// * [NormalizeMinMax](xref:Microsoft.ML.NormalizationCatalog.NormalizeMinMax(Microsoft.ML.TransformsCatalog, System.String, System.String, System.Int64, System.Boolean))
+    /// * [NormalizeMeanVariance](xref:Microsoft.ML.NormalizationCatalog.NormalizeMeanVariance(Microsoft.ML.TransformsCatalog,System.String,System.String,System.Int64,System.Boolean,System.Boolean))
+    /// * [NormalizeLogMeanVariance](xref:Microsoft.ML.NormalizationCatalog.NormalizeLogMeanVariance(Microsoft.ML.TransformsCatalog,System.String,System.String,System.Int64,System.Boolean))
+    /// * [NormalizeBinning](xref:Microsoft.ML.NormalizationCatalog.NormalizeBinning(Microsoft.ML.TransformsCatalog,System.String,System.String,System.Int64,System.Boolean,System.Int32))
+    /// * [NormalizeSupervisedBinning](xref:Microsoft.ML.NormalizationCatalog.NormalizeSupervisedBinning(Microsoft.ML.TransformsCatalog,System.String,System.String,System.String,System.Int64,System.Boolean,System.Int32,System.Int32))
+    /// ]]>
+    /// </format>
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Data/Transforms/Normalizer.cs]: can you move it outside of remarks section and just use seealso?

Comment (singlis) [reply src/Microsoft.ML.Data/Transforms/Normalizer.cs]: Unfortunately not as this hits assembly issues. I would have to move code which was deemed lower priority. So for now I did this to at least have some kind of reference.

File: src/Microsoft.ML.Data/Transforms/Normalizer.cs (hunk 4 above)
Comment (singlis) [reply src/Microsoft.ML.Data/Transforms/Normalizer.cs]: sorry I read too quickly through Shahab's original comment and thought he was asking if we supported integers. Yes we do support vectors, sorry for the confusion. I am updating.

---
In reply to: [277070189](https://github.com/dotnet/machinelearning/pull/3432#discussion_r277070189) [](ancestors = 277070189,277069180,277045796,276994571)

File: src/Microsoft.ML.Data/Transforms/Normalizer.cs (hunk 6 above)
Comment (singlis) [reply src/Microsoft.ML.Data/Transforms/Normalizer.cs]: Thanks Ivan - so if its 0,1 unless fixZero is set, why not then just document the normal case? 

---
In reply to: [277070876](https://github.com/dotnet/machinelearning/pull/3432#discussion_r277070876) [](ancestors = 277070876)

Comment (singlis) [reply src/Microsoft.ML.Data/Transforms/Normalizer.cs]: I have added further description to address this.


---
In reply to: [277084973](https://github.com/dotnet/machinelearning/pull/3432#discussion_r277084973) [](ancestors = 277084973,277070876)

File: src/Microsoft.ML.Data/Transforms/Normalizer.cs [hunk 8]
```diff
@@ -26,6 +26,43 @@
...
+    /// <see cref="IEstimator{TTransformer}"/> for the <see cref="NormalizingTransformer"/>.
+    /// </summary>
+    /// <remarks>
+    /// <format type="text/markdown"><![CDATA[
+    ///
+    /// ###  Estimator Characteristics
+    /// |  |  |
+    /// | -- | -- |
+    /// | Does this estimator need to look at the data to train its parameters? | Yes |
+    /// | Input column data type | <xref:System.Single> or<xref:System.Double> or a known-sized vector of those types. |
```
Comment (wschin) [inline src/Microsoft.ML.Data/Transforms/Normalizer.cs]:
```suggestion
    /// | Input column data type | <xref:System.Single> or <xref:System.Double> or a known-sized vector of those types. |
```

File: src/Microsoft.ML.Data/Transforms/Normalizer.cs [hunk 9]
```diff
@@ -26,6 +26,43 @@
...
+    /// | Output column data type | The same data type as the input column |
+    ///
+    /// The resulting NormalizingEstimator will normalize the data in one of the following ways based upon how it was created:
+    /// * Min Max - A linear rescale that is based upon the minimum and maximum values for each row.
+    /// * Mean Variance - Rescale each row to unit variance and, optionally, zero mean.
+    /// * Log Mean Variance - Rescale each row to unit variance based on a log scale.
+    /// * Binning - Bucketizes the data in each row and performs a linear rescale based on the calculated bins.
```
Comment (wschin) [inline src/Microsoft.ML.Data/Transforms/Normalizer.cs]: The explanation of binning doesn't carry much information. Also, we need equations for all of them.

Comment (singlis) [reply src/Microsoft.ML.Data/Transforms/Normalizer.cs]: Talked offline and agreed to address later. Having equations for all of our transformer/trainers is a much larger issue to address.

---
In reply to: [277110413](https://github.com/dotnet/machinelearning/pull/3432#discussion_r277110413) [](ancestors = 277110413)

File: src/Microsoft.ML.Data/Transforms/Normalizer.cs (hunk 8 above)
Comment (singlis) [reply src/Microsoft.ML.Data/Transforms/Normalizer.cs]: thank you for catching the space.

---
In reply to: [277110149](https://github.com/dotnet/machinelearning/pull/3432#discussion_r277110149) [](ancestors = 277110149)

**Pull request merged**